range-requests/src/lib.rs

```diff
@@ -0,0 +1,441 @@
+        res = res.status(StatusCode::OK);
+    }
+
+    res
+}
+
+/// Represents the raw, unparsed values of "range" from a request header.
+pub struct PotentialRange(Vec<u8>);
+
+impl PotentialRange {
+    /// Parses a single range request out of the range request.
+    ///
+    /// `len` is the total length of the document, for the range request being made.
+    ///
+    /// On failure, returns a range response with the appropriate headers
+    /// to inform the caller how to make a correct range request.
+    pub fn parse(&self, len: u64) -> Result<SingleRange, Response<Body>> {
+        self.single_range(len).map_err(|_err| bad_range_response(len))
+    }
+
+    fn single_range(&self, len: u64) -> Result<SingleRange, Error> {
+        match http_range::HttpRange::parse_bytes(&self.0, len) {
+            Ok(ranges) => {
+                if ranges.len() != 1 || ranges[0].length < 1 {
+                    // Right now, we don't want to deal with encoding a
+                    // response that has multiple ranges.
+                    Err(Error::MultipleRangesUnsupported)
+                } else {
+                    Ok(SingleRange::new(ranges[0], len)?)
+                }
+            }
+            Err(err) => Err(Error::Parse(err)),
+        }
+    }
+}
+
+/// A parsed range request, and associated "total document length".
+#[derive(Clone, Debug)]
+pub struct SingleRange {
+    range: http_range::HttpRange,
+    total: u64,
+}
+
+#[cfg(test)]
+impl PartialEq for SingleRange {
+    fn eq(&self, other: &Self) -> bool {
+        self.range.start == other.range.start
+            && self.range.length == other.range.length
+            && self.total == other.total
+    }
+}
+
+impl SingleRange {
+    pub fn new(
+        range: http_range::HttpRange,
+        total: u64,
+    ) -> Result<Self, Error> {
+        let http_range::HttpRange { start, mut length } = range;
+
+        const INVALID_RANGE: Error =
+            Error::Parse(http_range::HttpRangeParseError::InvalidRange);
+
+        // Clip the length to avoid going beyond the end of the total range
+        if start.checked_add(length).ok_or(INVALID_RANGE)? >= total {
+            length = total.checked_sub(start).ok_or(INVALID_RANGE)?;
+        }
+        // If the length is zero, we cannot satisfy the range request
+        if length == 0 {
+            return Err(INVALID_RANGE);
+        }
+
+        Ok(Self { range: http_range::HttpRange { start, length }, total })
+    }
+
+    /// Return the first byte in this range for use in inclusive ranges.
+    pub fn start(&self) -> u64 {
+        self.range.start
+    }
+
+    /// Return the last byte in this range for use in inclusive ranges.
+    pub fn end_inclusive(&self) -> u64 {
+        assert!(self.range.length > 0);
+
+        self.range
+            .start
+            .checked_add(self.range.length)
+            .unwrap()
+            .checked_sub(1)
+            .unwrap()
```

Comment:
nitpicky, goofy: perhaps these unwraps ought to be expects? i think both are more or less guaranteed unreachable, but it might be nice to have the expects document why?

Reply:
+1 on making these expects

Reply:
Updated!